tests/models/score_network/test_score_network_equivariance.py

```diff
@@ -0,0 +1,526 @@
+                for batch_idx in range(batch_size)
+            ]
+        )
+        return self.create_batch(
+            new_relative_coordinates,
+            new_cartesian_positions,
+            new_atom_types,
+            basis_vectors,
+            times,
+            noises,
+            forces,
+        )
+
+    def test_translation_invariance(self, output, translated_output):
+        torch.testing.assert_close(output, translated_output)
+
+    @pytest.fixture()
+    def rotated_scores_should_match(
+        self, is_rotations_cubic_point_group, is_cell_cubic, are_basis_vectors_rotated
+    ):
+        # The rotated scores should match the original scores if the basis vectors are rotated.
+        # If the basis vectors are NOT rotated, only a cubic unit cell (and cubic symmetries) should match.
+        should_match = are_basis_vectors_rotated or (
+            is_cell_cubic and is_rotations_cubic_point_group
+        )
+        return should_match
+
+    @pytest.fixture()
+    def atom_output_should_be_tested_for_rational_equivariance(self):
+        return True
+
+    def test_rotation_equivariance(
+        self,
+        output,
+        rotated_output,
+        basis_vectors,
+        rotated_basis_vectors,
+        cartesian_rotations,
+        rotated_scores_should_match,
+        atom_output_should_be_tested_for_rational_equivariance
+    ):
+
+        # The score is ~ nabla_x ln P. There must a be a basis change to turn it into a cartesian score of the
+        # form ~ nabla_r ln P.
+        reciprocal_basis_vectors = get_reciprocal_basis_vectors(basis_vectors)
+        cartesian_scores = einops.einsum(
+            reciprocal_basis_vectors,
+            output.X,
+            "batch alpha i, batch natoms i -> batch natoms alpha",
+        ).contiguous()
+
+        reciprocal_rotated_basis_vectors = get_reciprocal_basis_vectors(
+            rotated_basis_vectors
+        )
+        rotated_cartesian_scores = einops.einsum(
+            reciprocal_rotated_basis_vectors,
+            rotated_output.X,
+            "batch alpha i, batch natoms i -> batch natoms alpha",
+        ).contiguous()
+
+        expected_rotated_cartesian_scores = self.apply_rotation_to_configuration(
+            cartesian_rotations, cartesian_scores
+        )
+
+        if rotated_scores_should_match:
+            torch.testing.assert_close(
+                expected_rotated_cartesian_scores, rotated_cartesian_scores
+            )
+            torch.testing.assert_close(output.L, rotated_output.L)
+
+            if atom_output_should_be_tested_for_rational_equivariance:
+                torch.testing.assert_close(output.A, rotated_output.A)
+        else:
+            with pytest.raises(AssertionError):
+                torch.testing.assert_close(
+                    expected_rotated_cartesian_scores, rotated_cartesian_scores
+                )
+            # TODO: it's not clear what the expectation should be for A and L in this case...
```

Comment:
I am not sure, I would need to think more about it. It's fine to leave as a TODO for now

Reply:
Sounds good.